File: frontend/src/services/chatService.js
import apiClient from './api.js';

// Service de chat pour communiquer avec le backend
export const chatService = {
  // Envoyer un message
  async sendMessage(messageData) {
    try {
      const response = await apiClient.post('/chat/send', messageData);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de l\'envoi du message:', error);
      throw error;
    }
  },

  // Récupérer les messages d'une conversation
  async getConversationMessages(conversationId, page = 0, size = 50) {
    try {
      const response = await apiClient.get(`/chat/conversation/${conversationId}`, {
        params: { page, size }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des messages:', error);
      throw error;
    }
  },

  // Récupérer les messages entre deux utilisateurs
  async getMessagesBetweenUsers(userId1, userId2, page = 0, size = 50) {
    try {
      const response = await apiClient.get(`/chat/messages/${userId1}/${userId2}`, {
        params: { page, size }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des messages:', error);
      throw error;
    }
  },

  // Marquer les messages comme lus
  async markMessagesAsRead(receiverId, conversationId) {
    try {
      const response = await apiClient.put(`/chat/read/${receiverId}/${conversationId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors du marquage des messages:', error);
      throw error;
    }
  },

  // Récupérer les messages non lus
  async getUnreadMessages(receiverId) {
    try {
      const response = await apiClient.get(`/chat/unread/${receiverId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des messages non lus:', error);
      throw error;
    }
  },

  // Récupérer le nombre de messages non lus
  async getUnreadCount(receiverId) {
    try {
      const response = await apiClient.get(`/chat/unread-count/${receiverId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération du nombre de messages non lus:', error);
      throw error;
    }
  },

  // Récupérer les conversations d'un utilisateur
  async getUserConversations(userId) {
    try {
      const response = await apiClient.get(`/chat/conversations/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des conversations:', error);
      throw error;
    }
  },

  // Envoyer un message système
  async sendSystemMessage(messageData) {
    try {
      const response = await apiClient.post('/chat/system-message', messageData);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de l\'envoi du message système:', error);
      throw error;
    }
  },

  // Supprimer un message
  async deleteMessage(messageId) {
    try {
      const response = await apiClient.delete(`/chat/message/${messageId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la suppression du message:', error);
      throw error;
    }
  },

  // Modifier un message
  async editMessage(messageId, messageData) {
    try {
      const response = await apiClient.put(`/chat/message/${messageId}`, messageData);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la modification du message:', error);
      throw error;
    }
  },

  // Récupérer les statistiques du chat
  async getChatStatistics() {
    try {
      const response = await apiClient.get('/chat/statistics');
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des statistiques:', error);
      throw error;
    }
  },

  // Recherche avancée de messages
  async searchMessages(searchRequest) {
    try {
      const response = await apiClient.post('/chat/search/advanced', searchRequest);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche de messages:', error);
      throw error;
    }
  },

  // Recherche avec tolérance aux fautes de frappe
  async searchWithTypoTolerance(userId, searchTerm) {
    try {
      const response = await apiClient.get(`/chat/search/typo-tolerant/${userId}`, {
        params: { term: searchTerm }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche avec tolérance:', error);
      throw error;
    }
  },

  // Recherche par mots-clés multiples
  async searchByMultipleKeywords(userId, keywords) {
    try {
      const response = await apiClient.post(`/chat/search/keywords/${userId}`, { keywords });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche par mots-clés:', error);
      throw error;
    }
  },

  // Recherche de messages média
  async searchMediaMessages(userId) {
    try {
      const response = await apiClient.get(`/chat/search/media/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche de messages média:', error);
      throw error;
    }
  },

  // Recherche de messages système
  async searchSystemMessages(userId) {
    try {
      const response = await apiClient.get(`/chat/search/system/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche de messages système:', error);
      throw error;
    }
  },

  // Recherche par période
  async searchByPeriod(userId, startDate, endDate) {
    try {
      const response = await apiClient.get(`/chat/search/period/${userId}`, {
        params: { startDate, endDate }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche par période:', error);
      throw error;
    }
  },

  // Recherche par contenu
  async searchByContent(userId, content) {
    try {
      const response = await apiClient.get(`/chat/search/content/${userId}`, {
        params: { content }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche par contenu:', error);
      throw error;
    }
  },

  // Recherche par type de message
  async searchByMessageType(userId, messageType) {
    try {
      const response = await apiClient.get(`/chat/search/type/${userId}`, {
        params: { type: messageType }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche par type:', error);
      throw error;
    }
  },

  // Recherche de messages non lus
  async searchUnreadMessages(userId) {
    try {
      const response = await apiClient.get(`/chat/search/unread/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche de messages non lus:', error);
      throw error;
    }
  },

  // Recherche de messages modifiés
  async searchEditedMessages(userId) {
    try {
      const response = await apiClient.get(`/chat/search/edited/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche de messages modifiés:', error);
      throw error;
    }
  },

  // Recherche globale
  async searchGlobal(searchTerm) {
    try {
      const response = await apiClient.get('/chat/search/global', {
        params: { term: searchTerm }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la recherche globale:', error);
      throw error;
    }
  },

  // Obtenir des suggestions de recherche
  async getSearchSuggestions(searchTerm) {
    try {
      const response = await apiClient.get('/chat/search/suggestions', {
        params: { term: searchTerm }
      });
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des suggestions:', error);
      throw error;
    }
  },

  // Obtenir les statistiques de recherche d'un utilisateur
  async getSearchStatistics(userId) {
    try {
      const response = await apiClient.get(`/chat/search/statistics/${userId}`);
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la récupération des statistiques de recherche:', error);
      throw error;
    }
  },

  // Vérifier la santé du service de recherche
  async checkSearchHealth() {
    try {
      const response = await apiClient.get('/chat/search/health');
      return response.data;
    } catch (error) {
      console.error('❌ Erreur lors de la vérification de la santé du service:', error);
      throw error;
    }
  }
};

export default chatService;

File: src/main/java/com/diravenir/service/ChatService.java
package com.diravenir.service;

import com.diravenir.Entities.ChatMessage;
import com.diravenir.dto.ChatMessageDTO;
import com.diravenir.repository.ChatMessageRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class ChatService {
    
    private final ChatMessageRepository chatMessageRepository;
    
    /**
     * Envoyer un message
     */
    public ChatMessageDTO sendMessage(ChatMessageDTO messageDTO) {
        try {
            // Générer un ID de conversation si non fourni
            if (messageDTO.getConversationId() == null) {
                messageDTO.setConversationId(generateConversationId(messageDTO.getSenderId(), messageDTO.getReceiverId()));
            }
            
            ChatMessage message = messageDTO.toEntity();
            message.setSentAt(LocalDateTime.now());
            message.setIsRead(false);
            
            ChatMessage savedMessage = chatMessageRepository.save(message);
            log.info("Message envoyé: {} -> {}", messageDTO.getSenderId(), messageDTO.getReceiverId());
            
            return ChatMessageDTO.fromEntity(savedMessage);
        } catch (Exception e) {
            log.error("Erreur lors de l'envoi du message", e);
            throw new RuntimeException("Erreur lors de l'envoi du message", e);
        }
    }
    
    /**
     * Récupérer les messages d'une conversation
     */
    public List<ChatMessageDTO> getConversationMessages(String conversationId) {
        try {
            List<ChatMessage> messages = chatMessageRepository.findByConversationIdOrderBySentAtAsc(conversationId);
            return messages.stream()
                    .map(ChatMessageDTO::fromEntity)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Erreur lors de la récupération des messages de la conversation: {}", conversationId, e);
            throw new RuntimeException("Erreur lors de la récupération des messages", e);
        }
    }
    
    /**
     * Récupérer les messages entre deux utilisateurs
     */
    public List<ChatMessageDTO> getMessagesBetweenUsers(String userId1, String userId2) {
        try {
            List<ChatMessage> messages = chatMessageRepository.findMessagesBetweenUsers(userId1, userId2);
            return messages.stream()
                    .map(ChatMessageDTO::fromEntity)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Erreur lors de la récupération des messages entre utilisateurs: {} et {}", userId1, userId2, e);
            throw new RuntimeException("Erreur lors de la récupération des messages", e);
        }
    }
    
    /**
     * Marquer les messages comme lus
     */
    public void markMessagesAsRead(String receiverId, String conversationId) {
        try {
            chatMessageRepository.markMessagesAsRead(receiverId, conversationId);
            log.info("Messages marqués comme lus pour l'utilisateur: {} dans la conversation: {}", receiverId, conversationId);
        } catch (Exception e) {
            log.error("Erreur lors du marquage des messages comme lus", e);
            throw new RuntimeException("Erreur lors du marquage des messages", e);
        }
    }
    
    /**
     * Récupérer les messages non lus pour un utilisateur
     */
    public List<ChatMessageDTO> getUnreadMessages(String receiverId) {
        try {
            List<ChatMessage> messages = chatMessageRepository.findUnreadMessagesByReceiverId(receiverId);
            return messages.stream()
                    .map(ChatMessageDTO::fromEntity)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Erreur lors de la récupération des messages non lus pour: {}", receiverId, e);
            throw new RuntimeException("Erreur lors de la récupération des messages non lus", e);
        }
    }
    
    /**
     * Compter les messages non lus pour un utilisateur
     */
    public Long countUnreadMessages(String receiverId) {
        try {
            return chatMessageRepository.countUnreadMessagesByReceiverId(receiverId);
        } catch (Exception e) {
            log.error("Erreur lors du comptage des messages non lus pour: {}", receiverId, e);
            throw new RuntimeException("Erreur lors du comptage des messages non lus", e);
        }
    }
    
    /**
     * Récupérer les dernières conversations d'un utilisateur
     */
    public List<ChatMessageDTO> getLastMessagesByUserId(String userId) {
        try {
            List<ChatMessage> messages = chatMessageRepository.findLastMessagesByUserId(userId);
            return messages.stream()
                    .map(ChatMessageDTO::fromEntity)
                    .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Erreur lors de la récupération des dernières conversations pour: {}", userId, e);
            throw new RuntimeException("Erreur lors de la récupération des conversations", e);
        }
    }
    
    /**
     * Envoyer un message système
     */
    public ChatMessageDTO sendSystemMessage(String receiverId, String message) {
        try {
            ChatMessageDTO systemMessage = ChatMessageDTO.builder()
                    .senderId("SYSTEM")
                    .senderType(ChatMessage.SenderType.SYSTEM)
                    .receiverId(receiverId)
                    .receiverType(ChatMessage.SenderType.STUDENT)
                    .message(message)
                    .messageType(ChatMessage.MessageType.SYSTEM_MESSAGE)
                    .conversationId(generateConversationId("SYSTEM", receiverId))
                    .build();
            
            return sendMessage(systemMessage);
        } catch (Exception e) {
            log.error("Erreur lors de l'envoi du message système", e);
            throw new RuntimeException("Erreur lors de l'envoi du message système", e);
        }
    }
    
    /**
     * Supprimer un message
     */
    public void deleteMessage(Long messageId) {
        try {
            chatMessageRepository.deleteById(messageId);
            log.info("Message supprimé: {}", messageId);
        } catch (Exception e) {
            log.error("Erreur lors de la suppression du message: {}", messageId, e);
            throw new RuntimeException("Erreur lors de la suppression du message", e);
        }
    }
    
    /**
     * Générer un ID de conversation unique
     */
    private String generateConversationId(String userId1, String userId2) {
        // Trier les IDs pour assurer la cohérence
        String[] ids = {userId1, userId2};
        java.util.Arrays.sort(ids);
        return UUID.randomUUID().toString() + "_" + ids[0] + "_" + ids[1];
    }
    
    /**
     * Récupérer les statistiques de chat
     */
    public ChatStatistics getChatStatistics() {
        try {
            long totalMessages = chatMessageRepository.count();
            long unreadMessages = chatMessageRepository.countUnreadMessagesByReceiverId("ALL");
            
            return ChatStatistics.builder()
                    .totalMessages(totalMessages)
                    .unreadMessages(unreadMessages)
                    .build();
        } catch (Exception e) {
            log.error("Erreur lors de la récupération des statistiques de chat", e);
            throw new RuntimeException("Erreur lors de la récupération des statistiques", e);
        }
    }
    
    @lombok.Data
    @lombok.Builder
    public static class ChatStatistics {
        private Long totalMessages;
        private Long unreadMessages;
    }
}
